text/0000-century-of-the-self-type.md

```diff
@@ -0,0 +1,178 @@
+- Feature Name: arbitrary_self_types
+- Start Date: 2018-02-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow types that implement `Deref` targeting `Self` to be the receiver of a
+method. If the receiver type also implements the correct `CoerceUnsized` bound,
+that method is object safe.
+
+# Motivation
+[motivation]: #motivation
+
+Today, methods can only be received by value, by reference, by mutable
+reference, or by a `Box<Self>`. This has always intended to be generalized to
+support any kind of pointer, such as an `Rc<Self>` or an `Arc<Self>`. Since
+late 2017, it has been available on nightly under the `arbitrary_self_types`
+feature.
+
+This feature is increasingly relevant because of the role of special pointer
+types to constraint self-referential types, such as generators containing
+internal references. Because different kinds of "smart pointers" can constrain
+the semantics in non trivial ways, traits can rely on certain assumptions about
+the receiver of their method, whereas just implementing the trait *for* a smart
+pointer doesn't allow that kind of reliance.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to
+be any type `T` where `T: Deref<Target = Self>`. Shorthand exists, so that
+`self` with no ascription is of type `Self`, and `&self` and `&mut self` are of
+type `&Self` and `&mut Self`:
+
+```rust
+// All of these are valid:
+trait Foo {
+    fn by_value(self: Self);
+    fn by_ref(self: &Self);
+    fn by_ref_mut(self: &mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_arc(self: Arc<Self>);
+}
+```
+
+## Recursive arbitrary receivers
+
+Like the rule for deref coercions, the rule for receivers is recursive. If type
+`T` implements `Deref` targeting type `U`, and type `U` implements `Deref`
+targeting `Self`, `T` is a valid receiver (and so on outward).
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_ref_to_rc(self: &Rc<Self>) { ... }
+}
+```
+
+## Object safety
+
+In order for these receivers to be object safe, some additional traits need to
+be implemented. Given a reference type `Ptr<dyn Trait>`, the compiler must be
+able to prove that `T: Unsize<dyn Trait>` implies `Ptr<T>:
+CoerceUnsized<Ptr<dyn Trait>>`. If the compiler can prove this, methods with
+these receivers are object safe (how they object safe conversion is implemented
+is discussed later in the detailed design).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Handling trait objects
+
+To support object-safety, we must have a way of obtaining a vtable from the
+reference type, and then passing the correct receiver type to the function in 
+the vtable.
```

Comment:
Have you considered interactions with multiple trait bounds in the future?

Reply:
I don't think there's an interaction. This RFC doesn't introduce any constraint that doesn't already exist today.